feat(chunkserver): Make struct masterconn a class

lgsilva3087 opened the pull request:
This change prepares the masterconn module of the chunkserver to support multiple MDS connections in the future. The main changes are:

- MasterConn is now a class with more responsibilities, which fosters later support for multiple connections.
- Simpler callbacks handling from the global masterconn functions.
- Simplified polling.
- Small documentation (will continue).

Next commits will move the MasterConn class to own files, to keep simple the management of multiple connections in masterconn.h and masterconn.cc.

lgsilva3087 committed:
feat(chunkserver): Make struct masterconn a class

This change prepares the masterconn module of the chunkserver to
support multiple MDS connections in the future. The main changes are:

- MasterConn is now a class with more responsibilities, which fosters
  later support for multiple connections.
- Simpler callbacks handling from the global masterconn functions.
- Simplified polling.
- Small documentation (will continue).

Next commits will move the MasterConn class to own files, to keep
simple the management of multiple connections in masterconn.h and
masterconn.cc.

Signed-off-by: guillex <guillex@leil.io>

rolysr reviewed (approved):
LGTM.

ralcolea reviewed (approved):
Great work @lgsilva3087 :+1: :fire: :muscle:

lgsilva3087 committed:
feat(chunkserver): Use shared_ptr for gJobPool

gJobPool is meant to be shared between MasterConn instances.
This change makes it a shared_ptr and adds a jobPool_ member variable
to the MasterConn class, to safely protect the ownwership and lifetime.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
feat(chunkserver): Add own options to MasterConn

Previously, all MasterConn instances were using the same global
variables for the connection options (master host, master port, etc.).
This change adds own members to MasterConn objects to support multiple
connections in the future.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
feat(chunkserver): Decentralize MasterConn stats

The variables related to `bytes received` and `bytes sent` stats now
belong to the MasterConn objects which allows to know their values for
specific connections and reduces the number of global variables.

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
feat(chunkserver): Create JobPool before connections

The shared JobPool is created in advance before any MasterConn is
initiated. The MasterConn(s) are created now before registering the
callbacks in the event loop, which is safer.

Co-authored-by: Crash <crash@leil.io>

Signed-off-by: guillex <guillex@leil.io>

antuan96314 commented on src/chunkserver/init.h:
```diff
@@ -39,11 +39,11 @@ inline const std::vector<RunTab> runTabs = {
     RunTab{loadPlugins, "plugin manager"}, RunTab{hddInit, "hdd space manager"},
     // Has to be before "masterconn"
     RunTab{mainNetworkThreadInit, "main server module"},
+    RunTab{masterconn_init_threads, "master connection module - threads"},
     RunTab{masterconn_init, "master connection module"},
```
Consider merging this two init functions for simplicity.

lgsilva3087 replied:
I like the idea, but those functions will have more substance if we decide to add the second JobPool (for replications) and multiple MasterConn(s). I will keep them for now, with strong willing to unify them if possible.

dmga44 commented on src/chunkserver/masterconn.cc:
```diff
@@ -78,619 +91,731 @@ struct MasterConn {
-	if (eptr->mode == ConnectionMode::CONNECTED) {
+	if (eptr->mode() == ConnectionMode::CONNECTED) {
 		pdesc.emplace_back(gJobFD, POLLIN, 0);
 		gJobFDpDescPos = static_cast<int32_t>(pdesc.size() - 1);
```
Nice job here, though I want to point at a couple of things:
- the order of the two fd's to poll at are switched (it was previously the bgJobsPool listener and then the socket from master, and now it is the other way around) but this should not make any difference because the positions of each one are clear.
- as the bgJobsPool was kept global, then it means if there were several master connections, all of them were going there. Going a bit forward in that direction, it would be not easy to determine which master sent the job in the first place (in order to send the reply).

dmga44 replied:
These are just things to think about, are not worrying for now.

lgsilva3087 replied:
Order restored.

dmga44 commented on src/chunkserver/masterconn.cc:
```diff
@@ -700,86 +825,46 @@ void masterconn_send_status() {
-		} else {
-			eptr->mode = ConnectionMode::KILL;
-		}
+	
+	// Check if there are any background jobs to process.
+	if ((eptr->mode() == ConnectionMode::CONNECTED) && gJobFDpDescPos >= 0 &&
+	    (pdesc[gJobFDpDescPos].revents & POLLIN)) {
+		gJobPool->processCompletedJobs();
```
Please consider the following:
- in the previous implementation this piece of code was called after checking the status of the polling on the master connection fd, so it was clear that state of the connection was CONNECTED and this operation was safe (or safer).
- the proposed implementation does not has this kind of check, so the connection might died and needs to be tested (and reconnected if necessary). The issue comes when trying to reply to the master some of the tasks it sent, which could fail in this scenario.

lgsilva3087 replied:
Done.

dmga44 commented on src/chunkserver/masterconn.cc:
```diff
@@ -78,619 +91,731 @@ struct MasterConn {
 
-void masterconn_gotpacket(MasterConn *eptr, PacketHeader header, const MessageBuffer& message) try {
-	switch (header.type) {
+			// stats_bytesout += bytesWritten;
```
This seems to be some old code commented, consider removing it.

lgsilva3087 replied:
Removed.

dmga44 commented on src/chunkserver/masterconn.cc:
```diff
@@ -78,619 +91,731 @@ struct MasterConn {
+		if (jobPool_) {
+			job_create(
+			    *jobPool_, [this](uint8_t status, void *packet) { sauJobFinished(status, packet); },
+			    outputPacket, chunkId, chunkVersion, chunkType);
```
Consider the following:
- define the following funtion in the .cc (and .h for the purpose of the change):
```cpp
std::function<void(uint8_t status, void *packet)> sauJobFinished(void *eptr) {
	auto *masterConn = static_cast<MasterConn *>(eptr);
	return [masterConn](uint8_t status, void *packet) {
		masterConn->sauJobFinished(status, packet);
	};
}
```
- substitute each of the lambdas in the job calls by a call to that function, for instance:
```cpp
			job_create(*jobPool_, ::sauJobFinished(this), outputPacket, chunkId, chunkVersion,
			           chunkType);
```

This is just to make the code look better IMO, left to your consideration.

lgsilva3087 replied:
Done.

dmga44 reviewed (commented):
Overall, great job @lgsilva3087 🔥 !!

Please check out my comments, I think one of them is important.

As an extra minor suggestion, consider updating the logging to `safs::log_*` equivalent in the lines touched.

lgsilva3087 committed:
fix(chunkserver): Apply code review suggestions

Main changes:
- Poll descriptors order restored from previous version.
- Check for poll errors before trying to process jobs.
- New static sauJobFinished to simplify callback calling.

Co-authored-by: Dave <dave@leil.io>

Signed-off-by: guillex <guillex@leil.io>

lgsilva3087 committed:
chore(chunkserver): Use safs::log functions in masterconn

Old logging related safs_* calls were replaced by their equivalent
safs::log*.

Co-authored-by: Dave <dave@leil.io>

Signed-off-by: guillex <guillex@leil.io>

ralcolea commented on src/chunkserver/masterconn.cc:
```diff
@@ -78,620 +89,728 @@ struct MasterConn {
+			job_create(*jobPool_, sauJobFinished(this), outputPacket, chunkId, chunkVersion,
+			           chunkType);
+		} else {
+			safs::log_err("masterconn_create: jobPool is null.");
```
Consider updating the logs with the new function names for the functions involved in chunk operations.
For example, replace `"masterconn_create"` with `"createChunk"`.

lgsilva3087 replied:
Done.

ralcolea reviewed (approved):
Great job @lgsilva3087!!! :+1: :fire: :muscle:
